New Resource & Data Source: `azurerm_app_service_virtual_network_connection_gateway`

njuCZ opened the pull request:
Add new resource and data source for Gateway Required App Service Virtual Network Connection
(fix [#4333, #4332])

katbyte commented on examples/app-service/gateway-required-virtual-network-integration/README.md:
```diff
@@ -0,0 +1,9 @@
+### Notes
+
+* create a virtual network gateway may takes about 40 min.
+* To configure a gateway required virtual network integration, please make sure the virtual network already have a gateway configured with point-to-site addresses.
```
Can we not do this via terraform so the user doesn't have to?

njuCZ replied:
yes, user can use terraform to ceate a virtual network gateway configured with point-to-site addresses. in the example, I have shown that in the example tf files. 
In the notes, I want to mention the users how they can use the virtual network integration

katbyte replied:
Could we do everything required here in the example? it shouldn't require the user doing any setup aside from maybe changing some variables

katbyte commented on website/docs/d/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,77 @@
+data "azurerm_app_service_virtual_network_connection" "example" {
+  resource_group_name   = "example-resource-group"
+  app_service_name      = "example-appservice"
+  vnet_name             = "example-virtual-network"
```
Could we order this alphabetically?

katbyte commented on website/docs/d/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,77 @@
+
+* `app_service_name` - (Required) Specifies the name of the App Service.
+
+* `resource_group_name` - (Required) The name of the resource group which the app service belongs to.
```
COuld we use data source phrasing:
`pecifies the name of the Resource Group where the PostgreSQL Server exists.`

katbyte commented on website/docs/d/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,77 @@
+
+* `resource_group_name` - (Required) The name of the resource group which the app service belongs to.
+
+* `vnet_name` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.
```
This is not a location? could we use the correct description here?

katbyte commented on website/docs/d/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,77 @@
+* `name` - The name of this route. This is only returned by the server and does not need to be set by the client.
+* `start_address` The starting address for this route. This may also include a CIDR notation, in which case the end address must not be specified.
+*  `end_address` - The ending address for this route. If the start address is specified in CIDR notation, this must be omitted.
+---
```
Could we use data source phrasing and grammer here?

katbyte commented on website/docs/d/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,77 @@
+
+* `type` - the TYPE of this resource
+
+* `properties` - A `properties` block as defined below.
```
We shoukldn't have a properties block here

katbyte commented on website/docs/d/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,77 @@
+
+* `id` - The ID of the App Service Virtual Network Connection.
+
+* `name` - The NAME of the App Service Virtual Network Connection.
```
capitalization error here

katbyte commented on website/docs/d/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,77 @@
+
+* `name` - The NAME of the App Service Virtual Network Connection.
+
+* `type` - the TYPE of this resource
```
COuld we elaborate on this? what does type mean? also TYPE is capitalized

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,169 @@
+# azurerm_app_service_virtual_network_connection
+
+Manages Gateway Required Virtual Network Connection (Integrate with an App Service).
+This function is Based on script https://gallery.technet.microsoft.com/scriptcenter/Connect-an-app-in-Azure-ab7527e3
```
I don't think we need this?

katbyte replied:
Same comment, could we remove this?

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
````diff
@@ -0,0 +1,169 @@
+## Example Usage
+
+```hcl
+resource "azurerm_resource_group" "test" {
````
Could we use example for all resource identifiers ?
```suggestion
resource "azurerm_resource_group" "example" {
```

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
````diff
@@ -0,0 +1,169 @@
+
+```hcl
+resource "azurerm_resource_group" "test" {
+  name     = "test"
````
Could we not use `test` for names in examples?

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,169 @@
+This function is Based on script https://gallery.technet.microsoft.com/scriptcenter/Connect-an-app-in-Azure-ab7527e3
+
+To configure a gateway required virtual network integration, please make sure the virtual network already have a gateway configured with point-to-site addresses.
+Please refer to this doc: https://docs.microsoft.com/en-us/azure/app-service/web-sites-integrate-with-vnet#gateway-required-vnet-integration
```
Can this not be set via terraform?

katbyte replied:
Would we show the terraform required to do this in the example and maybe make this a note?

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,169 @@
+
+* `app_service_name` - (Required) Specifies the name of the App Service.
+
+* `resource_group_name` - (Required) The name of the resource group which the app service belongs to.
```
Could we use resource phrasing here? 
`The name of the resource group in which the Automation Account is created.`

katbyte replied:
should this also be force new?

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,169 @@
+
+* `resource_group_name` - (Required) The name of the resource group which the app service belongs to.
+
+* `kind` - (Optional) Kind of resource.
```
Could we elaborate on what this is and possible values?

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,169 @@
+`properties` supports the following:
+
+* `vnet_resource_id` - (Required) The Virtual Network's resource ID.
+* `dns_servers` - (Optional) DNS servers to be used by this Virtual Network. This should be a comma-separated list of IP addresses.
```
We shouldn't use comma seperate list here, it should be a `TypeList`

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,169 @@
+
+* `type` - the TYPE of this resource
+
+* `properties` - A `properties` block as defined below.
```
this seems duplicated?

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,169 @@
+* `kind` - Kind of resource.
+* `name` - Resource Name.
+* `type` - Resource type.
+* `properties` - A `route properties` block as defined below.
```
Could we elaborate on these descriptions and use full sentences?

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,169 @@
+* `dns_servers` - DNS servers to be used by this Virtual Network. This should be a comma-separated list of IP addresses.
+* `resync_required` - true if a resync is required; otherwise, false.
+* `vnet_resource_id` - The Virtual Network's resource ID.
+* `routes` - (Array) One or more `route` block defined below.
```
Could we elaborate on these descriptions and use full sentences

katbyte commented on azurerm/resource_arm_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,513 @@
+		}
+	}
+
+	if err := checkAppServiceExist(client, ctx, &resGroup, &appServiceName); err != nil {
```
Why are we checking for the existence of these resources instead of letting the API do it?

katbyte commented on azurerm/resource_arm_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,513 @@
+	}
+	if virtualNetworkGateway.VpnClientConfiguration == nil || virtualNetworkGateway.VpnClientConfiguration.VpnClientAddressPool == nil {
+		return fmt.Errorf("this gateways %q under vnet %q (Resource Group %q) does not have a Point-to-site Address Range. Please specify one in CIDR notation, e.g. 10.0.0.0/8", gatewayName, vnetName, vnetResGroup)
+	}
```
Why is this required? shouldn't the API handle this?

njuCZ commented on azurerm/resource_arm_app_service_virtual_network_connection_test.go:
```diff
@@ -0,0 +1,124 @@
+resource "azurerm_resource_group" "testvnet" {
+  name     = "acctestRG-%d"
+  location = "%s"
+}
```
I want to show how to integrate an app service in one region with virtual network in another region.
It's OK to show the test case in one region

njuCZ commented:
> Hi @njuCZ,
> 
> I started reviewing this and have left a bunch of comments inline. While going through the create function it appears we are calling a bunch of different APIs. Would this resource make more sense a a few individual ones that combine for the same affect? possible a `virtual_network_gateway_connection` and a `virtual_network_gateway_connection_app_service_association` resource?

this resource is to intergrate an app service with a virtual network. It's one whole function in the azure portal, though it needs call two different api. there is no need to separate into two different resources

njuCZ commented:
@katbyte I have modified my codes according to your suggestion, could you kindly review my codes once more?

katbyte commented on azurerm/data_source_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,150 @@
+			"virtual_network_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Could we validate this with at least `NoEmptyStrings` (and we might have an existing vnet name validator from that resource?)

katbyte commented on azurerm/data_source_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,150 @@
+	if props := resp.VnetInfoProperties; props != nil {
+		if props.VnetResourceID != nil {
+			d.Set("virtual_network_id", props.VnetResourceID)
+		}
```
we don't need to check for nil here as `d.Set` will do it for us:
```
d.Set("virtual_network_id", props.VnetResourceID)
```

katbyte commented on azurerm/data_source_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,150 @@
+
+		if props.DNSServers != nil {
+			d.Set("dns_servers", props.DNSServers)
+		}
```
Same here

katbyte commented on azurerm/data_source_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,150 @@
+			d.Set("dns_servers", props.DNSServers)
+		}
+
+		if routes := props.Routes; routes != nil {
```
Could we move the nil check inside the flatten function?

katbyte commented on azurerm/resource_arm_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,400 @@
+		return err
+	}
+
+	ipConfigurationId, err := azure.ParseAzureResourceID(*(*gatewaySubnet.IPConfigurations)[0].ID)
```
I think we'll need to check these dereferences & len != 0?

katbyte commented on azurerm/resource_arm_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,400 @@
+		}
+		if props.DNSServers != nil {
+			d.Set("dns_servers", props.DNSServers)
+		}
```
As mentioned above we don't need to check for nil as d.Set will

katbyte commented on azurerm/resource_arm_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,400 @@
+		if props.DNSServers != nil {
+			d.Set("dns_servers", props.DNSServers)
+		}
+		if routes := props.Routes; routes != nil {
```
And the nil check should go in the flatten function

katbyte commented on azurerm/resource_arm_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,400 @@
+	}
+
+	for _, subnet := range *respVnet.Subnets {
+		if *subnet.Name == "GatewaySubnet" {
```
Could we nil check here?

katbyte commented on azurerm/resource_arm_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,400 @@
+		attr["name"] = *route.Name
+		attr["start_address"] = *route.VnetRouteProperties.StartAddress
+		attr["end_address"] = *route.VnetRouteProperties.EndAddress
+		attr["route_type"] = string(route.VnetRouteProperties.RouteType)
```
We'll need to nil check all these derefences

katbyte commented on azurerm/resource_arm_app_service_virtual_network_connection_test.go:
```diff
@@ -0,0 +1,115 @@
+func testAccAzureRMAppServiceVirtualNetworkConnection_basic(rInt int, location string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "example" {
+	name     = "acctestRG-%d"
```
could we use 
```suggestion
	name     = "acctestRG-appservice-%d"
```

here please?

katbyte commented on website/docs/d/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,69 @@
+
+* `app_service_name` - (Required) Specifies the name of the App Service.
+
+* `resource_group_name` - (Required) The name of the resource group which the app service belongs to.
```
could we reorder this
```
* `resource_group_name` - (Required) The name of the resource group which the app service belongs to.

* `app_service_name` - (Required) Specifies the name of the App Service.
```

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,146 @@
+
+* `app_service_name` - (Required) Specifies the name of the App Service.
+
+* `resource_group_name` - (Required) The name of the resource group which the app service belongs to.
```
could we change the ordering here too?

katbyte commented:
Also we are getting test failures:
```
------- Stdout: -------
=== RUN   TestAccDataSourceAzureRMAppServiceVirtualNetworkConnection_basic
=== PAUSE TestAccDataSourceAzureRMAppServiceVirtualNetworkConnection_basic
=== CONT  TestAccDataSourceAzureRMAppServiceVirtualNetworkConnection_basic
--- FAIL: TestAccDataSourceAzureRMAppServiceVirtualNetworkConnection_basic (2946.44s)
    testing.go:569: Step 0 error: errors during apply:
        
        Error: error add certificate for gateway "example": error adding cerfiticate for gateway "example" (Resource Group "acctestRG-191114233037998613"): network.VirtualNetworkGatewaysClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code="AnotherOperationInProgress" Message="Another operation on this or dependent resource is in progress. To retrieve status of the operation use uri: https://management.azure.com/subscriptions/1a6092a6-137e-4025-9a7c-ef77f76f2c02/providers/Microsoft.Network/locations/westeurope/operations/a672b118-fb92-4de0-911a-2bdd5a3c3050?api-version=2019-06-01." Details=[]
```

njuCZ commented:
@katbyte thank you for your suggestion, I have fixed them

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,145 @@
+
+* `dns_servers` - DNS servers to be used by this Virtual Network. This should be a comma-separated list of IP addresses.
+
+* `resync_required` - true if a resync is required; otherwise, false.
```
Could we make this a proper sentence?

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,145 @@
+
+* `certificate_thumbprint` - The client certificate thumbprint.
+
+* `dns_servers` - DNS servers to be used by this Virtual Network. This should be a comma-separated list of IP addresses.
```
This should be a TypeList

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,145 @@
+page_title: "Azure Resource Manager: azurerm_app_service_virtual_network_connection"
+sidebar_current: "docs-azurerm-resource-app-service-virtual-network-connection"
+description: |-
+  Manages Gateway Required Virtual Network Connection (Integrate with an App Service).
```
These descriptions should be:
```suggestion
  Manages an App Service Virtual Network Connection
```

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,145 @@
+
+  depends_on = [
+    azurerm_virtual_network_gateway.example,
+    azurerm_app_service.example
```
We shouldn't need this?

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,145 @@
+  virtual_network_id    = "${azurerm_virtual_network.example.id}"
+
+  depends_on = [
+    azurerm_virtual_network_gateway.example,
```
If this is always required would it make sense to use the `azurerm_virtual_network_gateway` id above so the dependency is enforced by the schema?

katbyte commented on azurerm/resource_arm_app_service_virtual_network_connection.go:
```diff
@@ -0,0 +1,403 @@
+			},
+
+			"dns_servers": {
+				Type:     schema.TypeString,
```
I think this should be a TypeList?

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,145 @@
+resource "azurerm_app_service_virtual_network_connection" "example" {
+  app_service_name      = "${azurerm_app_service.example.name}"
+  resource_group_name   = "${azurerm_resource_group.example.name}"
+  virtual_network_id    = "${azurerm_virtual_network.example.id}"
```
if we make this `azurerm_virtual_network_gateway_id` we won't need the depends on. I presume you can't create this resource if it doesn't exist then?

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,145 @@
+* `name` - Resource Name.
+* `route_type` - The type of route this is: DEFAULT - By default, every app has routes to the local address ranges specified by RFC1918 INHERITED - Routes inherited from the real Virtual Network routes STATIC - Static route set on the app only. Valid values are `DEFAULT`, `INHERITED`, `STATIC`
+* `start_address` The starting address for this route. This may also include a CIDR notation, in which case the end address must not be specified.
+* `end_address` - The ending address for this route. If the start address is specified in CIDR notation, this must be omitted.
```
These are exported so i dont' think we need 
```suggestion
* `end_address` - The ending address for the route.
```

katbyte commented on website/docs/r/app_service_virtual_network_connection.html.markdown:
```diff
@@ -0,0 +1,145 @@
+* `id` - Resource Id.
+* `name` - Resource Name.
+* `route_type` - The type of route this is: DEFAULT - By default, every app has routes to the local address ranges specified by RFC1918 INHERITED - Routes inherited from the real Virtual Network routes STATIC - Static route set on the app only. Valid values are `DEFAULT`, `INHERITED`, `STATIC`
+* `start_address` The starting address for this route. This may also include a CIDR notation, in which case the end address must not be specified.
```
These are exported so i dont' think we need 
```suggestion
* `start_address` The starting address for this route.
```

njuCZ commented:
sorry for late reply, and I have modified this PR according to the suggestions.

the createUpdate process is complex, the basic flow is:
1. CreateOrUpdateVnetConnection
2. update virtual network gateway using the cert info of step 1's result
3. get vpnPackageURL from virtual network gateway
4. CreateOrUpdateVnetConnectionGateway using step 3's result

because I think step1 and step 4 are not actually a resource, so I wirte these in one resource. But it needs to update virtual network gateway, So I added `ignore changes` in test case(or I can set this field in virtual network gateway `Optional:true` `Computed: true`? ),  do you think it's OK? Do I still need to split them into two resource ?

njuCZ commented:
hi @katbyte I have refactored to `azurerm_app_service_virtual_network_connection_gateway`

katbyte commented on azurerm/internal/services/web/data_source_app_service_virtual_network_connection_gateway.go:
```diff
@@ -0,0 +1,143 @@
+		return fmt.Errorf("Error making Read request on App Service Virtual Network Connection for app %q vnet %q (Resource Group %q): %+v", appServiceName, vnetName, resGroup, err)
+	}
+
+	d.SetId(*resp.ID)
```
Could we nil check respo.ID and for ""?

katbyte commented on azurerm/internal/services/web/resource_arm_app_service_virtual_network_connection_gateway.go:
```diff
@@ -0,0 +1,405 @@
+		return err
+	}
+	vnetResGroup := id.ResourceGroup
+	vnetName := id.Path["virtualNetworks"]
```
Could we write a parse function and move to it like in #5779 ?

katbyte commented on azurerm/internal/services/web/tests/resource_arm_app_service_virtual_network_connection_gateway_test.go:
```diff
@@ -0,0 +1,114 @@
+
+func testAccAzureRMAppServiceVirtualNetworkConnectionGateway_basic(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
```
We'll need update tests to have provider blocks now

katbyte commented on azurerm/internal/services/web/tests/resource_arm_app_service_virtual_network_connection_gateway_test.go:
```diff
@@ -0,0 +1,114 @@
+
+  lifecycle {
+    ignore_changes = [
+      vpn_client_configuration.0.root_certificate,
```
Why are we ignoring lifecycle changes on this property? it might want to be computed if its being updated externally now

njuCZ replied:
Because the steps are 
1. CreateOrUpdateVnetConnection
2. result of step 1 contains cert infomation, we should set the cert to virtual network gateway (check duplicate)
3. generate vpn package uri
4. CreateOrUpdateVnetConnectionGateway using step 3's result

a gateway required virtual network integration needs these four steps. In this process, we need update certificate of the virtual network gateway.
I think there is no means to split this process.

I have added `computed: true` in the newest codes for virtual network gateway

katbyte commented on examples/app-service/gateway-required-virtual-network-integration/main.tf:
```diff
@@ -0,0 +1,79 @@
+resource "azurerm_resource_group" "example" {
```
We'll need a provider block here

katbyte commented on website/docs/d/app_service_virtual_network_connection_gateway.html.markdown:
```diff
@@ -0,0 +1,66 @@
+page_title: "Azure Resource Manager: azurerm_app_service_virtual_network_connection_gateway"
+sidebar_current: "docs-azurerm-resource-app-service-virtual-network-connection-gateway"
+description: |-
+  Get Gateway Required Virtual Network Connection (Integrate with an App Service).
```
SHould this be
```suggestion
  Get an App Service Virtual Network Connection Gateway.

```

katbyte commented on website/docs/d/app_service_virtual_network_connection_gateway.html.markdown:
```diff
@@ -0,0 +1,66 @@
+
+# Data Source: azurerm_app_service_virtual_network_connection_gateway
+
+Use this data source to access information about an existing App Service integrated Gateway Required Virtual Network Connection.
```
Wording seems off here

katbyte commented on website/docs/d/app_service_virtual_network_connection_gateway.html.markdown:
```diff
@@ -0,0 +1,66 @@
+A `route` block supports the following:
+* `name` - Resource Name.
+* `route_type` - The type of route this is: DEFAULT - By default, every app has routes to the local address ranges specified by RFC1918 INHERITED - Routes inherited from the real Virtual Network routes STATIC - Static route set on the app only. Valid values are `DEFAULT`, `INHERITED`, `STATIC`
+* `start_address` The starting address for this route. This may also include a CIDR notation, in which case the end address must not be specified.
```
Could we change the wording here to match existing data sources?

katbyte commented on website/docs/r/app_service_virtual_network_connection_gateway.html.markdown:
```diff
@@ -0,0 +1,148 @@
+page_title: "Azure Resource Manager: azurerm_app_service_virtual_network_connection_gateway"
+sidebar_current: "docs-azurerm-resource-app-service-virtual-network-connection-gateway"
+description: |-
+  Manages an Gateway Required App Service Virtual Network Connection
```
as above

katbyte commented on website/docs/r/app_service_virtual_network_connection_gateway.html.markdown:
```diff
@@ -0,0 +1,148 @@
+
+# azurerm_app_service_virtual_network_connection_gateway
+
+Manages Gateway Required Virtual Network Connection (Integrate with an App Service).
```
as above

katbyte commented on website/docs/r/app_service_virtual_network_connection_gateway.html.markdown:
```diff
@@ -0,0 +1,148 @@
+
+* `dns_servers` - DNS servers to be used by this Virtual Network. It is a list of IP addresses.
+
+* `resync_required` - is resync required
```
```suggestion
* `resync_required` - Is resync required?

```

katbyte commented on website/docs/r/app_service_virtual_network_connection_gateway.html.markdown:
```diff
@@ -0,0 +1,148 @@
+* `name` - Resource Name.
+* `route_type` - The type of route this is: DEFAULT - By default, every app has routes to the local address ranges specified by RFC1918 INHERITED - Routes inherited from the real Virtual Network routes STATIC - Static route set on the app only. Valid values are `DEFAULT`, `INHERITED`, `STATIC`
+* `start_address` The starting address for the route.
+* `end_address` - The ending address for the route.
```
these are exported so could we word them as such?

katbyte commented on website/docs/r/app_service_virtual_network_connection_gateway.html.markdown:
```diff
@@ -0,0 +1,148 @@
+* `start_address` The starting address for the route.
+* `end_address` - The ending address for the route.
+---
+
```
Could we include the timeouts docs here?

njuCZ commented on azurerm/internal/services/web/resource_arm_app_service_virtual_network_connection_gateway.go:
```diff
@@ -0,0 +1,405 @@
+		return fmt.Errorf("this gateways %q under vnet %q (Resource Group %q) does not have a Point-to-site Address Range. Please specify one in CIDR notation, e.g. 10.0.0.0/8", gatewayName, vnetName, vnetResGroup)
+	}
+
+	isRelated, err := checkGatewayInVirtualNetwork(virtualNetworkGateway, vnetId)
```
I have left some comments in the codes for the steps. At first I want to find the virtual network gateway from the virtual network, now we have both parameters: `virtual_network_id` and `virtual_network_gateway_id`, so I first check whether they are related. (thus virtual network gateway should located in a `gateway` subnet in the virtual network)

njuCZ commented:
@katbyte sorry for late reply. I have refined my codes.
The logic is a little complex for this function, you can mainly refer to this script
https://gallery.technet.microsoft.com/scriptcenter/Connect-an-app-in-Azure-ab7527e3

Thanks again for your patience reviewing this PR!

jackofallops commented:
Hi @njuCZ 
I've been working through this PR to see how we can get it progressed. 

It feels like things are confused (and confusing) as there should be two resources (with data sources) here, `azurerm_app_service_network_connection` and `azurerm_app_service_network_connection_gateway`?

I think a redesign into 2 will simplify things, give better focus to each part, and help us get this through. WDYT?

tombuildsstuff commented:
Based [on this comment](https://github.com/terraform-providers/terraform-provider-azurerm/pull/4458#issuecomment-603946426) there's some design questions which need to be addressed here - rather than leaving this open whilst that's happening I'm going to close this PR for the moment, but once those have been resolved we can circle back around and take another look here
